Add support for a per-stream version of the "Default code block programming language".

kuv2707 opened the pull request:
This PR adds support for a per-channel version of the default code block language, following a similar addition to Realm.
- If the stream default language is not set, the realm default is used.
- For non-channel messages, the realm default code block language is used.

Fixes: #17454 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
<details>
<summary>
Screenshots and screen captures:
</summary>

![image](https://github.com/zulip/zulip/assets/88523649/4446e325-4471-4610-b84e-34240a046636)



</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 committed:
compose_state: Add function to retrieve current stream.

This change has been made as a prep for #17454, where we
need to access the properties of the currently selected stream
in the compose box preview and the typeahead.

kuv2707 commented:
Ready for a review!

pratik-pc commented:
[Screencast from 08-06-24 02:23:43 AM IST.webm](https://github.com/zulip/zulip/assets/33497322/3bbf81fe-edcd-4c78-a16c-1e2265bd1e54)

I am not able to change the codeblock language

kuv2707 commented:
@pratik-pc There was a minor bug (missed a line during rebasing). Please have a look now.

pratik-pc commented:
[Screencast from 10-06-24 02:12:01 PM IST.webm](https://github.com/zulip/zulip/assets/33497322/5762b303-4a77-4179-95b2-7a274e3d58be)

@kuv2707 discarding `dropdown-widget` changes doesn't reset the dropdown state as it does with all other dropdowns. Also, the spacing between the `dropdown-widget` seems a bit off at first glance.

kuv2707 commented:
Thanks for pointing that out!
I had missed it

kuv2707 commented:
@pratik-pc fixed it!

pratik-pc commented:
[Screencast from 11-06-24 01:09:25 PM IST.webm](https://github.com/zulip/zulip/assets/33497322/e1eca43d-6f86-4727-86e5-95184037411d)

@kuv2707 The `save` option doesn't show up when the `code block language` dropdown is saved to a different option and again changed to the previous one.

kuv2707 commented:
@pratik-pc thanks for pointing that out, fixed it (it was introduced after the recent conversion of the `settings_components` module to typescript)

pratik-pc commented:
> discarding dropdown-widget changes doesn't reset the dropdown state as it does with all other dropdowns. Also, the spacing between the dropdown-widget seems a bit off at first glance.

> The save option doesn't show up when the code block language dropdown is saved to a different option and again changed to the previous one.

Highlighting this manual test fixes for next review stage. In addition, also tested in narrow window and if default language for the stream changes successfully. Changes in dropdown reflect correctly now after testing.

Also, realm default code language is not reflected in stream modal. (Don't know the intended design for this)

@kuv2707 It would be better if 'Fixes part of' were included instead of 'Fixes' in all but the last commit message. Other than that, this looks good to me

pratik-pc commented:
@zulipbot remove "buddy review"

kuv2707 commented:
@zulipbot add "mentor review"

madrix01 commented on zerver/models/streams.py:
```diff
@@ -34,6 +34,7 @@ class Stream(models.Model):
     deactivated = models.BooleanField(default=False)
     description = models.CharField(max_length=MAX_DESCRIPTION_LENGTH, default="")
     rendered_description = models.TextField(default="")
+    stream_default_code_block_language = models.TextField(default="")
```
should we rename this to `default_code_block_language`?

kuv2707 replied:
I had been thinking so

madrix01 commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -435,7 +435,7 @@ def run(self, lines: Iterable[str]) -> List[str]:
         default_language = None
         if isinstance(self.md, ZulipMarkdown) and self.md.zulip_realm is not None:
-            default_language = self.md.zulip_realm.default_code_block_language
+            default_language = self.md.default_code_block_language
```
Instead of passing value here can we do computation at view level can do computation here? Like pass `zulip_stream` along with `zulip_realm` and do something like,

```py
default_language = None
if isinstance(self.md, ZulipMarkdown) and self.md.zulip_realm and self.md.zulip_stream:
    if self.md.zulip_stream.default_code_block_language:
        default_language = self.md.zulip_stream.default_code_block_language
    elif self.md.zulip_realm.default_code_block_language:
        default_language = self.md.zulip_realm.default_code_block_language

```

madrix01 replied:
This would look clean and you'll not have to pass the `default_code_block_language` parameter everywhere. This is a suggesstion, please do tell if there is any other way. @kuv2707 @Riken-Shah

Riken-Shah replied:
Quickly skimming through the code, The only problem with this idea is limiting future reusability. If we refactor the code and move the logic for selecting `default_code_block_langauge` here, we might have to rewrite the logic. 

Delegating this logic to the lowest level ([`do_convert`](https://github.com/zulip/zulip/pull/30208/files/8a62d7adef0e966946c6676d48849b0dd38fa660#diff-fdf42dee6384fb054fbc027e0b7b4f68896624321adb1b5f7a4ccf89094b92edR2632-R2634)) does sound correct IMO. 

Would love to know your thoughts @kuv2707.

kuv2707 replied:
I had tried something similar.
We would already have access to the message's stream through the message object itself. But the problem came during rendering previews of the messages. 
This is handled by the function `render_messages_backend`. It only receives the message content and the user profile, and manually creates a dummy `Message` object containing the bare minimum required to make this code path work. We don't have access to the stream of the message during rendering previews, so if we decide to include the stream object at the lowest level, we'd need to make a database query for the entire stream object, only to then retrieve its `default_code_block_language`. This seemed unnecessary, so I ended up directly passing the language in the `render_message_markdown` function and using it down to the lowest level.

kuv2707 replied:
```
class ZulipMarkdown(markdown.Markdown):
    zulip_message: Optional[Message]
    zulip_realm: Optional[Realm]
    zulip_db_data: Optional[DbData]
    zulip_rendering_result: MessageRenderingResult
    image_preview_enabled: bool
    url_embed_preview_enabled: bool
    url_embed_data: Optional[Dict[str, Optional[UrlEmbedData]]]
    default_code_block_language: str
    ```
    
    This class already seems to have some options related to the message and its rendering, so I thought it would be best to append the default language property to it.

alya replied:
@Riken-Shah do you have additional feedback here?

Riken-Shah replied:
Nope, The @kuv2707 approach seems correct!

kuv2707 commented:
@madrix01 @Riken-Shah  I have addressed the feedback. Please have a look.

madrix01 commented:
@kuv2707 have gone through the PR, impletation looks good to me. Please do rebase the PR, looks like there are merge conflicts. 
cc @Riken-Shah

kuv2707 commented:
@zulipbot remove "mentor review"

@zulipbot  add "maintainer review"

alya commented:
@kuv2707 You can ping @sahil839 for a review once this PR has been rebased.

kuv2707 commented:
@sahil839 Please have a look.

I'll add the `changelog.md` entry once everything else is fine

sahil839 reviewed (commented):
Posted some comments. Haven't tested yet.

kuv2707 commented:
@sahil839 Please have a look.
I've addressed the previous feedback.

sahil839 reviewed (commented):
Posted some comments.

kuv2707 commented on zerver/openapi/zulip.yaml:
```diff
@@ -11580,6 +11582,8 @@ paths:
                 contentType: application/json
               folder_id:
                 contentType: application/json
+              default_code_block_language:
```
Oh, removed.

kuv2707 commented on version.py:
```diff
@@ -34,7 +34,7 @@
 # entries in the endpoint's documentation in `zulip.yaml`.
 
-API_FEATURE_LEVEL = 390
+API_FEATURE_LEVEL = 391
```
Yes, removed.

kuv2707 commented on api_docs/unmerged.d/ZF-991763.md:
Added change info.

kuv2707 commented on zerver/views/streams.py:
```diff
@@ -280,6 +281,7 @@ def update_stream_backend(
     can_remove_subscribers_group: Json[GroupSettingChangeRequest] | None = None,
     can_subscribe_group: Json[GroupSettingChangeRequest] | None = None,
     folder_id: Json[int | None] | MissingType = Missing,
+    default_code_block_language: str | None = None,
```
I had lost that in a faulty rebase.
Added it.

kuv2707 commented on zerver/lib/streams.py:
```diff
@@ -310,6 +314,7 @@ def create_stream_if_needed(
             is_in_zephyr_realm=realm.is_zephyr_mirror_realm,
             message_retention_days=message_retention_days,
             folder=folder,
+            default_code_block_language=default_code_block_language,
```
Fixed.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -21715,6 +21737,15 @@ paths:
+                    **Changes**: New in Zulip 11.0 (feature level ZF-991763).
+                  type: string
+                  example: python
+
```
We don't need this empty line.

sahil839 commented on zerver/tests/test_subs.py:
```diff
@@ -2999,6 +2999,20 @@ def test_permission_to_change_stream_folder(self) -> None:
+            {"default_code_block_language": "python"},
+        )
+        stream = get_stream("test_stream", realm)
+        self.assertEqual(stream.default_code_block_language, "python")
```
I meant testing it when user has permission through `can_administer_channel_group` like we do for other stream related settings and permissions.

kuv2707 replied:
Hmm ok, I've added a test case for that, following the pattern of a bunch of tests using `do_change_stream_group_based_setting`.

sahil839 commented on zerver/actions/message_send.py:
```diff
@@ -623,6 +625,12 @@ def build_message_send_dict(
+    if stream is not None:
+        # We need to use the default code block language of the stream
+        # for rendering the message.
+        default_code_block_language = stream.default_code_block_language
```
Would be good to add a test for this checking which language is used to render code blocks when sending messages to streams with different value set and when sending DMs.

sahil839 reviewed (commented):
Posted some comments.

kuv2707 commented on api_docs/unmerged.d/ZF-991763.md:
```diff
@@ -9,4 +9,3 @@
   to Stream and Subscription objects.
 * `POST /messages/render`: Added `default_code_block_language`
   which will be used when rendering code blocks in a stream message.
-  
```
Done.

sahil839 commented on zerver/worker/embed_links.py:
```diff
@@ -62,6 +62,7 @@ def consume(self, event: Mapping[str, Any]) -> None:
                 message.content,
                 realm,
                 url_embed_data=url_embed_data,
+                default_code_block_language="",
```
I think user would expect the stream level language to be used for code blocks irrespective of whether the message has embedded links or not.

kuv2707 replied:
Hmm, makes sense.
I've appended `default_code_block_language` to the event payload in `message_send.py` and `message_edit.py`.

kuv2707 commented on web/templates/stream_settings/stream_permissions.hbs:
```diff
@@ -215,6 +215,11 @@
+                {{> ../dropdown_widget_with_label
+                  widget_name=default_code_block_language_widget_name
+                  label="Default code block language"
+                  value_type="string"
```
Started [#frontend > Place for a new channel setting @ 💬](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Place.20for.20a.20new.20channel.20setting/near/2247973)

kuv2707 commented on web/src/settings_components.ts:
```diff
@@ -2017,3 +2024,65 @@ export function get_channel_folder_value_from_dropdown_widget($elem: JQuery): nu
+                const $subsection = stream_settings_containers
+                    .get_edit_container(sub)
+                    .find(".advanced-configurations-container");
+                save_discard_stream_settings_widget_status_handler($subsection, sub);
```
I ended up doing away with `get_edit_container` and followed the same pattern as `create_stream_group_setting_widget` - that of doing `$container.closest(".settings-subsection-parent")`.

kuv2707 committed:
stream_settings: Add dropdown widget for property.

Added dropdown widgets in stream create and
stream edit forms to modify the stream property
`default_code_block_language`.
We also handle server events on this field.

Fixes #17454

sahil839 commented on zerver/tests/test_message_send.py:
```diff
@@ -3718,3 +3719,31 @@ def test_message_send_in_channel_with_topics_disabled(self) -> None:
+        # If the stream level value is empty, then it should use the realm level value.
+        do_change_default_code_block_language(stream, "", acting_user=iago)
+        ret = check_message(iago, client, addressee, message_content, realm)
+        self.assertTrue('data-code-language="JavaScript"' in (ret.message.rendered_content or ""))
```
I guess you might have used `self.assertIsNotNone`. Adding `asert ret.message.rendered_content is not None` before `self.assertTrue` line should work.

kuv2707 replied:
Hmm that works, thanks.

kuv2707 commented:
@sahil839 I've made the suggested changes.

kuv2707 commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -1473,3 +1473,45 @@ def test_can_set_delete_message_policy_group(self) -> None:
+        )
+        self.assert_json_success(result)
+        stream = get_stream("test_stream", realm)
+        self.assertEqual(stream.default_code_block_language, "python")
```
Oh got it.
I moved this back to `test_subs.py` under class `SubscriptinAPITest`.
I've also removed the block checking permission since it would now be redundant.

kuv2707 commented on zerver/tests/test_link_embed.py:
````diff
@@ -1082,3 +1091,51 @@ def test_custom_title_replaces_youtube_url_title(self) -> None:
+                denmark_stream.name,
+                topic_name="editing",
+                content=f"```\nprint('Hello, World!')\n``` \n\n {url}",
+            )
````
Wouldn't it be better to test that in `test_message_sent.py`?

kuv2707 commented on zerver/tests/test_message_send.py:
```diff
@@ -3718,3 +3719,47 @@ def test_message_send_in_channel_with_topics_disabled(self) -> None:
+        )
+        self.assert_json_success(edit_result)
+        ret.message.refresh_from_db()
+        self.assertIn('data-code-language="Python"', ret.message.rendered_content)
```
Shifted it there, and added a few more test cases. I had to copy the setup code from here.

kuv2707 committed:
stream_settings: Use `default_code_block_language`.

For syntax highlighting in code blocks, the
`default_code_block_language` of the stream is used in stream
messages. If this property is not set, we fall back to the realm
default code block language.
For direct messages, only the realm default is used.

- Added the field to the Stream type definition.
- Modified the messages/render handler to accept a default
code block language for rendering a message.

Moving the message to a new channel doesn't change the
code block language of the message, but further edits
to the message content will use the new channel's language
setting (or fall back to realm default).

Fixes part of #17454.

sahil839 commented on zerver/actions/message_edit.py:
```diff
@@ -1576,6 +1576,10 @@ def check_update_message(
 
+    default_code_block_language = ""
+    if isinstance(message_edit_request, StreamMessageEditRequest):
+        default_code_block_language = message_edit_request.target_stream.default_code_block_language
```
Ok. I am not sure what should be the expected behavior. Would be good to discuss on CZO once.

kuv2707 replied:
Asked [#frontend > Default code block language on message move @ 💬](https://chat.zulip.org/#narrow/channel/6-frontend/topic/Default.20code.20block.20language.20on.20message.20move/near/2251776)

kuv2707 commented on zerver/tests/test_subs.py:
```diff
@@ -5182,6 +5182,39 @@ def test_notification_bot_dm_on_subscription(self) -> None:
         data = self.assert_json_success(response)
         self.assertNotIn("new_subscription_messages_sent", data)
 
+    def test_can_change_default_code_block_language(self) -> None:
```
But your [earlier comment](https://github.com/zulip/zulip/pull/30208#discussion_r2122883163) suggested to have this group test.

kuv2707 committed:
stream: Add field default_code_block_language.

The field has been added to the Stream model and related
models. Following the realm level field, the default value
is empty string.

This field stores the default language used in markdown
code blocks in messages sent to a channel, when none is specified.

Fixes part of #17454

sahil839 commented:
I have pushed back to the PR with the changes requested above. Also added some tests for completion and made a couple of other minor changes as well.

I have asked a couple of questions on ([#design > Default code block language on message move @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Default.20code.20block.20language.20on.20message.20move/near/2253310)) regarding rerendering of message changing the code block language, so PR would need to updated once we decide something on that.
